---
title: "IsoformSwitchAnalzyeR and IRFinder analysis"
author: "Stuart Lee, Charity Law"
date: "`r Sys.Date()`"
output: 
  bookdown::html_document2:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
                      fig.align = "center",
                      comment = "#>", 
                      message = FALSE,
                      warning = FALSE,
                      fig.path = here::here("img/"),
                      cache.path = here::here("cache/")
                      )

coverage_plot <- function(.y) {
  target <- filter(parts, gene_id == !!.y$gene_id)
  
  # set up parts
  cvg_rng <- as(lapply(cvg, function(x) join_parts(x, target)), "GRangesList")
  md <- DataFrame(Sample = Rle(names(cvg_rng), lengths(cvg_rng)),
                  CellLine = Rle(design$CellLine, lengths(cvg_rng)),
                  Kit = Rle(as.factor(design$Kit), lengths(cvg_rng)))
  cvg_rng <- unlist(cvg_rng, use.names = FALSE)
  mcols(cvg_rng) <- cbind(mcols(cvg_rng), md)
  cvg_rng <- cvg_rng %>% 
        mutate(log_score = log2(score + 0.5),
               strand = strand(target),
               var = paste0("Kit: ", Kit, " Cellline: ", CellLine))
  
  
  p <- view_coverage(cvg_rng, 
                     score = log_score, 
                     colour = feature_type,  
                     facets = dplyr::vars(var)) +
    ylim(0, NA) +
    guides(colour = FALSE) +
    scale_color_brewer(palette = "Dark2") +
    labs(subtitle = paste("Coverage over", target$gene_name),
         y = "Log coverage"
    ) +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          strip.text = element_text(hjust = 0, size = 8),
          strip.background = element_blank()) +
    expand_limits(y = 0)
  
  segments <- superintronic:::flatten_parts(target)
  track <- view_segments(segments, color = feature_type) +
    theme(axis.text.x = element_text(size = 8))

  patchwork::wrap_plots(p, track, 
                        ncol = 1, 
                        heights = c(2, 0.25))
  
} 
library(plyranges)
library(superintronic)

# external inputs
design <- readRDS(here::here("data", "design.rds"))
parts <- readRDS(here::here("data", "complete-annotation.rds"))
cvg <- readRDS(here::here("data", "complete-coverage.rds"))
all_features <- readRDS(here::here("data", "superintronic-features.rds"))
hits <- readRDS(here::here("data", "superintronic-hits.rds")) 

# update for relevant samples
design <- subset(design, Kit == "polyA")
cvg <- cvg[design$Sample]
```

##  IsoformSwitchAnalyzeR analysis

The Bioconductor package `IsoformSwitchAnalyzeR` also allows for the exploration
of intron retention events using a similar workflow:

1. import isoform quantifications from kallisto/salmon/other program of your choice.

We used kallisto quant on single read mode with an average fragment
size of 175 and a standard deviation of 20, on all fastq files for each
replicate in the polyA kits for each of our cellines. 
We need to use both here since 
IsoformSwitchAnalzyeR requires biological replicates for testing for
switching events.

```{r isf-prepare}
library(IsoformSwitchAnalyzeR)

quants <- importIsoformExpression(
  parentDir = here::here("data", "kallisto/"),
)

```


2. Preparing data structure called a `switchList` by importing annotations
and design matrix.

```{r isf-import}
# use a simplified version of the design
design_isf <- data.frame(sampleID = colnames(quants$abundance[,-1])) %>% 
  mutate(condition =  design$CellLine)

isf <- importRdata(
  isoformCountMatrix = quants$counts,
  isoformRepExpression = quants$abundance,
  designMatrix = design_isf,
  isoformExonAnnoation = here::here("data-raw", "gencode.v27.annotation.gtf.gz"),
  isoformNtFasta = here::here("data-raw", "gencode.v27.transcripts.fa.gz"),
  removeNonConvensionalChr = TRUE,
)

isf 
```


3. Filtering isoforms


Next we prefilter, the isoforms using roughly the same filters we used
in the superintronic analysis above. To speed things up we could further
specify a gene expression filter, but we have left that out for now.

```{r isf-filtered, cahce = TRUE, dependson=isf}
isf_filtered <- preFilter(isf,
                          acceptedGeneBiotype = "protein_coding",
                          removeSingleIsoformGenes = TRUE,
                          geneExpressionCutoff = 1,
                          isoformExpressionCutoff = 0.1,
                          keepIsoformInAllConditions = TRUE
                          )
isf_filtered
```

4. Searching for intron retention events

First we are required to run DEXSeq to test for isoform switching (
this does all pairwise comparisons between conditions, which slows things
down quite a bit.) and between the two kits. We have used quite a small
alpha since DEXSeq seems quite liberal...

```{r isf-dexseq}
isf_analysed <- isoformSwitchTestDEXSeq(isf_filtered, 
                                        reduceToSwitchingGenes = TRUE,
                                        alpha = 0.01,
                                        dIFcutoff = 0.1)
isf_analysed
```


Now we can analyse intron retention events from one kit relative to another:

```{r isf-intron-retention, cache=TRUE}
# this takes quite a while to compute...
isf_analysed <- analyzeIntronRetention(isf_analysed)
isf_analysed <- analyzeSwitchConsequences(isf_analysed, "intron_retention")

# the result
isf_analysed

# isoforms with intron retention events
isf_analysed$switchConsequence %>% 
  dplyr::count(switchConsequence)
```

And the results can be summarised at the gene level and compared to the 
superintronic analysis. 

At the gene level there are around 210 genes declared significant
at 0.01 fdr cut-off between the celllines kits that
are filtered for intron retention events.

```{r top-table}
is_res <- extractTopSwitches(isf_analysed, 
                             filterForConsequences = TRUE, 
                             n = NA) %>% 
  dplyr::as_tibble()

is_res
```

We can also look at the coverage plots for these genes (we will choose
the top 10 genes):

```{r isf-coverage-plots}
top10 <- is_res %>% 
  filter(Rank < 10)

for (i in seq_len(nrow(top10))) {
  input <- top10[i,]
  cplot <- coverage_plot(input)
  fname <- here::here("img", "isa-cov", 
                      paste0(input$gene_id, "-rank-", input$Rank, ".pdf"))
  ggsave(fname, cplot)
}
```


Comparing to the previous analysis with superintronic, there is one gene
overlap between the two methods. The comparison to superintronic is not
quite the same since this method is based on isoforms within genes between 
celllines rather
than our summaries over genes within each cellline separately.

```{r overlaps}
# only polyA results
polyA_sets <- hits %>%
  filter(Kit == "polyA") %>% 
  dplyr::distinct(gene_id) %>%
  arrange(gene_id) %>% 
  mutate(value = 1, 
         var = paste0("superintronic_", gsub("-", "", CellLine),"_", Kit)) %>% 
  ungroup() %>%  
  plyranges::select(gene_id, var, value) %>% 
  tidyr::spread(var, value) %>% 
  dplyr::mutate_at(2:3, .funs = ~ dplyr::if_else(is.na(.), 0L, 1L)) 


all_sets <- is_res %>%
  mutate(isa_set = 1L) %>% 
  select(gene_id, isa_set, gene_name, gene_switch_q_value, rank = Rank) %>% 
  dplyr::full_join(polyA_sets) %>% 
  dplyr::mutate_at(dplyr::vars(isa_set, 
                               dplyr::starts_with("superintronic")), 
                   .funs = ~ dplyr::if_else(is.na(.), 0L, .)) 
  
library(UpSetR)
plot_sets <- all_sets %>% 
  select(`IsoformSwitchAnalyzeR` = isa_set, 
         dplyr::starts_with("superintronic")) %>% 
  as.data.frame()
pdf(here::here("img", "olaps-superintronic-isa.pdf"))
upset(plot_sets)
dev.off()
```



### Comparison to IRFinder


```{r, echo = FALSE}
targets <- filter(parts, 
                  gene_name %in% c("NBEAL2", "HNRNPL", "HLA-B")
) %>% 
  select(gene_id) %>% 
  as.data.frame()

fig6_plots <- lapply(seq_len(nrow(targets)),
                function(.x) coverage_plot(targets[.x,]))

fig6 <- patchwork::wrap_plots(fig6_plots, ncol = 1, guides = "keep")

ggsave(here::here("figures/Fig6.pdf"), fig6, height = 7, width = 12)

```