---
title: "IsoformSwitchAnalzyeR analysis"
author: "Stuart Lee, Charity Law"
date: "`r Sys.Date()`"
output: 
  bookdown::html_document2:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
                      fig.align = "center",
                      comment = "#>", 
                      message = FALSE,
                      warning = FALSE,
                      fig.path = here::here("img/"),
                      cache.path = here::here("cache/")
                      )

save_plot_isr <- function(.x, .y) {
  target <- filter(parts, gene_id == !!.y$gene_id)
  cvg_sub <- join_parts(cvg, target)
  fname <- paste0(target$gene_name, "_isr_coverage_plot.pdf")
  fname <- here::here("img", fname)
  p <- view_coverage(cvg_sub, target, facets = "Kit")
  print(p)
  ggplot2::ggsave(fname, p, dpi = "retina")
}

# external inputs
cvg <- readr::read_rds(here::here("data", "cvg.rds"))
all_features <- readr::read_rds(here::here("data", "superintronic-features.rds"))
hits <- readr::read_rds(here::here("data", "superintronic-hits.rds"))
```

##  IsoformSwitchAnalyzeR analysis

The Bioconductor package `IsoformSwitchAnalyzeR` also allows for the exploration
of intron retention events using a similar workflow:

1. import isoform quantifications from kallisto/salmon/other program of your choice.

We used kallisto quant on single read mode with an average fragment
size of 200 and a standard deviation of 20, on all fastq files for each
replicate in the polyA/total RNA kits. We need to use both here since 
IsoformSwitchAnalzyeR requires biological replicates for testing for
switching events.

```{r isf-prepare}
library(IsoformSwitchAnalyzeR)

quants <- importIsoformExpression(
  parentDir = here::here("data", "kallisto/"),
)

```


2. Preparing data structure called a `switchList` by importing annotations
and design matrix.

```{r isf-import}
# use a simplified version of the design
design_isf <- data.frame(sampleID = colnames(quants$abundance[,-1])) %>% 
  mutate(condition =  rep(c("polyA", "total-RNA"), 3),
         replicate = rep(c("R1", "R2", "R3"), each = 2)
  )

isf <- importRdata(
  isoformCountMatrix = quants$counts,
  isoformRepExpression = quants$abundance,
  designMatrix = design_isf,
  isoformExonAnnoation = here::here("data-raw", "gencode.v27.annotation.gtf.gz"),
  isoformNtFasta = here::here("data-raw", "gencode.v27.transcripts.fa.gz"),
  removeNonConvensionalChr = TRUE,
)

isf 
```


3. Filtering isoforms


Next we prefilter, the isoforms using roughly the same filters we used
in the superintronic analysis above. To speed things up we could further
specify a gene expression filter, but we have left that out for now.

```{r isf-filtered, cahce = TRUE, dependson=isf}
isf_filtered <- preFilter(isf,
                          acceptedGeneBiotype = "protein_coding",
                          removeSingleIsoformGenes = TRUE,
                          geneExpressionCutoff = 1,
                          isoformExpressionCutoff = 0.1,
                          keepIsoformInAllConditions = TRUE
                          )
isf_filtered
```

4. Searching for intron retention events

First we are required to run DEXSeq to test for isoform switching (
this does all pairwise comparisons between conditions, which slows things
down quite a bit.) and between the two kits. We have used quite a small
alpha since DEXSeq seems quite liberal...

```{r isf-dexseq}
isf_analysed <- isoformSwitchTestDEXSeq(isf_filtered, 
                                        reduceToSwitchingGenes = TRUE,
                                        alpha = 0.01,
                                        dIFcutoff = 0.15)
isf_analysed
```


Now we can analyse intron retention events from one kit relative to another:

```{r isf-intron-retention, cache=TRUE}
# this takes quite a while to compute...
isf_analysed <- analyzeIntronRetention(isf_analysed)
isf_analysed <- analyzeSwitchConsequences(isf_analysed, "intron_retention")

# the result
isf_analysed

# isoforms with intron retention events
isf_analysed$switchConsequence %>% 
  dplyr::count(switchConsequence)
```

And the results can be summarised at the gene level and compared to the 
superintronic analysis. 

At the gene level there are around 502 genes declared significant
at 0.01 fdr cut-off between the polyA and total-RNA kits that
are filtered for intron retention events.

```{r top-table}
is_res <- extractTopSwitches(isf_analysed, 
                             filterForConsequences = TRUE, 
                             n = NA) %>% 
  dplyr::as_tibble()

is_res
```

We can also look at the coverage plots for these genes (we will choose
the top 20 genes):

```{r isf-coverage-plots}
is_res %>% 
  filter(Rank < 40) %>%
  group_by(gene_id) %>% 
  dplyr::group_walk(save_plot_isr)
```


Comparing to the previous analysis with superintronic, there is no
overlap between the two methods. (Of the genes in the IsoformSwitchAnalyzeR analysis, only
136 were kept by the filters in the superintronic, none of these were
selected by our rangenostics). The comparison to superintronic is not
quite the same since this method is based on isoforms within genes between 
kits rather
than our summaries over genes within just the polyA kit, 
however there still seems to be a lot of intron retention events by IsoformSwitchAnalyzeR.

```{r overlaps}
is_res <- is_res %>%
  dplyr::left_join(all_features %>% 
                     mutate(selected = gene_id %in% hits$gene_id) %>% 
                     dplyr::select(gene_id, selected))

dplyr::count(is_res, selected)
```