---
title: "Overlaps with index classifications"
author: "Stuart Lee"
date: "`r Sys.Date()`"
output: 
  bookdown::html_document2:
    toc: true
    number_sections: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      message = FALSE,
                      warning = FALSE,
                      fig.path = here::here("img/index_classifications/"))
library(plyranges)
parts <- readRDS(here::here("data", "complete-annotation.rds"))

hits <- readRDS(here::here("data", "superintronic-hits.rds"))

parts_tbl <- parts %>% 
  select(gene_id, gene_name, .drop_ranges = TRUE) %>% 
  tibble::as_tibble()

source(here::here("R/prettycov.R"))
```

## Extract superintronic hits 


* superintronic run on all kits in each of the HCC827 and NCIH1975 celllines
```{r si-sets}
# only polyA results
library(dplyr)
polyA_sets <- hits %>%
  filter(Kit == "polyA") %>% 
  distinct(gene_id) %>%
  arrange(gene_id) %>% 
  mutate(value = 1, 
         var = paste0("superintronic:", gsub("-", "", CellLine),"_", Kit)) %>% 
  ungroup() %>%  
  plyranges::select(gene_id, var, value) %>% 
  tidyr::spread(var, value) %>% 
  dplyr::mutate_at(2:3, .funs = ~ dplyr::if_else(is.na(.), 0L, 1L)) %>% 
  left_join(parts_tbl)

totalRNA_sets <- hits %>%
  filter(Kit == "total-RNA") %>% 
  distinct(gene_id) %>%
  arrange(gene_id) %>% 
  mutate(value = 1, 
         var = paste0("superintronic:", gsub("-", "", CellLine),"_", Kit)) %>% 
  ungroup() %>%  
  plyranges::select(gene_id, var, value) %>% 
  tidyr::spread(var, value) %>% 
  dplyr::mutate_at(2:3, .funs = ~ dplyr::if_else(is.na(.), 0L, 1L)) %>% 
  left_join(parts_tbl)
```


## Run index

```{r}
library(index)
exon <- readRDS(system.file("extdata/exon_dge.Rds", package = "index"))
intron <- readRDS(system.file("extdata/intron_dge.Rds", package = "index"))
group <- readRDS(system.file("extdata/group.Rds", package = "index"))

x <- index_analysis(exon, intron, group, p.value = 0.01)

res <- x$tops$intron %>% 
  mutate(classification = x$category) %>% 
  tibble::as_tibble()

index_res <- res %>% 
  filter(adj.P.Val < 0.01) %>% 
  select(gene_id = GeneID, index_pval = adj.P.Val, logfc = logFC) %>% 
  mutate(`index` = 1L) 

index_res
```

Using index, we specify a cut-off on differential intron counts at an 
FDR of 0.01. 


## index sets overlapping total RNA comparisons

All hits discovered by superintronic except for 19 genes are 
obtained by the index analysis using the beforementioned FDR threshold. 

```{r index_olap}
library(UpSetR)
dex_set <- index_res %>% 
  full_join(totalRNA_sets) %>% 
  mutate_at(vars(index, starts_with("superintronic")), 
            .funs = ~ if_else(is.na(.), 0L, .)) 

dex_upset <- upset(as.data.frame(dex_set))
dex_upset
```

```{r index-upset-1, include = FALSE}
pdf(here::here("img", "index_classifications", "index-superintronic-olaps-fdr_0.01.pdf"))
print(dex_upset)
dev.off()
```

How many genes in DEList overlap with HCC287 total RNA = 13 
How many genes in DEList overlap with NCIH1975 total RNA = 16
How many genes overlap between the overlapping genes = 12.  

If we look at the direction of the log fold-change  when there is overlap, the
superintronic results are concordant (the  index run has contrasts where
upregulated is comparing NCIH1975 to HCC827)

```{r}
dex_set %>% 
  count(`superintronic:HCC287_total-RNA`, 
        `superintronic:NCIH1975_total-RNA`, 
        index_sign = sign(logfc)) %>% 
  filter(`superintronic:HCC287_total-RNA` != 0 | 
           `superintronic:NCIH1975_total-RNA` != 0,
         !is.na(index_sign))
```

## Coverage plots for intron +/-

Next we look at the index category for `intron+-`, we first produce coverage
plots for the top 10 genes, ordered by t-statistic values. We do this over the
full design, instead of the just the total-RNA samples.

```{r intron-plus-minus}
# sort by t-statistic
intron_plus_minus <- res %>% 
  filter(grepl("Intron", classification), adj.P.Val < 0.01) %>%
  arrange(desc(abs(t)))

top10_genes <- intron_plus_minus %>% 
  dplyr::slice(1:10) %>% 
  select(gene_id =  GeneID, classification) %>% 
  mutate_all(as.character)

top10_genes
```

```{r read-rds, echo = FALSE}
cvg_over_features <- readRDS(here::here("data","complete-coverage.rds")) 
design <- readRDS(here::here("data", "design.rds"))
```

```{r cov-plots, warning = FALSE, echo = FALSE, message = FALSE, fig.width = 8}
library(ggplot2)
for (i in seq_len(nrow(top10_genes))) {
  current <-  top10_genes[i,]
  track_plot <- pretty_cov_plot(cvg_over_features, 
                                parts,
                                current,
                                design,
                                heights = c(4, 1/4))
  print(track_plot)
  out_name <- paste0(current$gene_id, "-", 
                     tolower(current$classification),
                     "_index_coverage.pdf")
  ggsave(here::here("img",
                    "index_classifications", out_name), 
         track_plot,
         width = 8,
         height = 12,
         units = "in")
}
```



## Overlaps with index classifications

Next we look at the overlaps between superintronic results
and all the  index classifications. Wen see there is only
gene classified by superintronic, that doesn't come up as a hit
in any of the index categories.  

```{r classification-set, echo = FALSE}
complete_set <- res %>%
  select(gene_id = GeneID, 
         index_pval = adj.P.Val, 
         logfc = logFC, 
         classification) %>% 
  filter(nchar(classification) > 0) %>% 
  mutate(index = 1L) %>% 
  tidyr::pivot_wider(names_from = classification,
                     values_from = index) %>%
  full_join(totalRNA_sets) %>%
  # rename(`HCC287 total RNA` = `superintronic:HCC287_total-RNA`,
  #        `NCIH1975 total RNA` = `superintronic:NCIH1975_total-RNA`) %>% 
  mutate_at(vars(-gene_id:-logfc, -gene_name),
            .funs = ~ if_else(is.na(.), 0L, .))

all_classification_upset <- upset(as.data.frame(complete_set),
                                  nsets = 10,
                                  text.scale = 0.7)
all_classification_upset
```

The remaining results make sense, most superintronic results fall into either `+` or `-` categories
a few fall into the remaining categories.

```{r index-upset-2, include = FALSE}
pdf(here::here("img", "index_classifications", "index-categories-superintronic-olaps.pdf"))
print(all_classification_upset)
dev.off()
```


Below we show a table, with the proportion of superintronic hits
for each CellLine within each category. In the tables, missing
values indicate the result was unique to superintronic.

These are the proportions for HCC287 cellline hits:

```{r, echo = FALSE}
by_category <- res %>%
  select(gene_id = GeneID, 
         index_pval = adj.P.Val, 
         logfc = logFC, 
         classification) %>% 
  filter(nchar(classification) > 0) %>% 
    full_join(totalRNA_sets) %>%
  mutate_at(vars(-gene_id:-classification, -gene_name),
            .funs = ~ if_else(is.na(.), 0L, .)) %>% 
  group_by(classification)

by_category %>% 
  filter(`superintronic:HCC287_total-RNA` == 1) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(proportion = n / sum(n)) %>% 
  arrange(desc(proportion))
```
Likewise, the NCIH1975 results make sense, we see that again the results make sense, 
where most classifications fall into plus categories.

```{r, echo = FALSE}
by_category %>% 
  filter(`superintronic:NCIH1975_total-RNA` == 1) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(proportion = n / sum(n)) %>% 
  arrange(desc(proportion))
```

Likewise, the proportions for NCIH1975 hits, show that most hits
fall into the plus category, which makes sense. 


## Coverage over shared intron +/- hits


```{r, echo = FALSE}
si_olaps <- complete_set %>% 
  filter(`Intron+` == 1 | `Intron-` == 1, 
         `superintronic:HCC287_total-RNA` == 1, 
         `superintronic:NCIH1975_total-RNA` == 1) %>% 
  select(gene_id, gene_name, starts_with("Intron")) %>% 
  tidyr::pivot_longer(cols = -gene_id:-gene_name, 
                      names_to = "classification") %>% 
  filter(value == 1)

for (i in seq_len(nrow(si_olaps))) {
  current <-  si_olaps[i,]
  track_plot <- pretty_cov_plot(cvg_over_features, 
                                parts,
                                current,
                                design,
                                heights = c(4, 1/4))
  print(track_plot)
  out_name <- paste0(current$gene_id, "-", 
                     tolower(current$classification),
                     "_index_coverage.pdf")
  ggsave(here::here("img", "index_classifications", out_name), 
         track_plot,
         width = 8,
         height = 12,
         units = "in")
}
```

## index overlaps with known NMD targets

These are from the following paper,

Colombo, M., Karousis, E. D., Bourquin, J., Bruggmann, R., & Mühlemann, O. (2017). Transcriptome-wide identification of NMD-targeted human mRNAs reveals extensive redundancy between SMG6- and SMG7-mediated degradation pathways. RNA , 23(2), 189–201. https://doi.org/10.1261/rna.059055.116


The NMD targets were derived in the following way via several knock-down experiments:

> Our final goal is to provide a unique score for every gene, to estimate its likelihood of being an NMD target. The first step to achieve such synthetic result was to combine KD and rescue conditions for every factor. A joint gene-specific log2FC value was generated by computing the average between the KD/Ctrl log2FC and the inverse of the rescue/KD log2FC. So if a gene is up-regulated in UPF1 KD compared to Ctrl and it is down-regulated in the rescue compared to the KD, it will have a high positive combined log2FC. We called this quantity KD-rescue log2FC. The significance of this combined log2FC was computed by a technique called sum of P-value (Edgington 1972). All genes down-regulated in a KD or up-regulated in a rescue, were assigned a P-value of 1 before applying this algorithm. We called this quantity KD-rescue P-value. In all cases in which we will refer to the significance or the log2FC of a single condition, like UPF1 or dKD_SMG6, we refer to these meta-analysis computations. Next, we aimed at finding the genes that complied with our definition of NMD target: gene reacting to UPF1 and at least one between SMG6 and SMG7. We therefore combined all conditions in a single list of significant results, using a set of P-value meta-analysis methods (Supplemental Table S2). The results from SMG6 and dKD_SMG6 were combined with Fisher's method in a single meta_SMG6 score. The same comparison was done for SMG7. A meta_SMGs significance score was then computed with a sum of P-value from meta_SMG6 and meta_SMG7. The final significance parameter used to determine the list of most significant NMD targets was calculated with a Fisher's method from meta_SMGs and UPF1_FDR (meta_meta).

We read in supplementary material via their provided excel spreadsheet
and restrict down to the gene_id and their final P-value used to select
targets.

```{r nmd-reads, warning = FALSE}
library(readxl)
nmd_targets <- read_xlsx(
  here::here("data", "colombo-nmd-targets.xlsx"), 
  skip = 5 
) %>% 
  select(gene_id = gene, adj_pvalue = meta_meta) 
```
We count how many genes there were at a P-value threshold of 0.01

```{r}
count(nmd_targets, adj_pvalue < 0.01)
```

We then take these results and produce a similar UpSet plot
to what we had before, combining the index classifications:
```{r, echo = TRUE}
index_hits <-res %>%
  select(gene_id = GeneID, 
         index_pval = adj.P.Val, 
         logfc = logFC, 
         classification) %>% 
  filter(nchar(classification) > 0) %>% 
  mutate(index = 1L) %>% 
  tidyr::pivot_wider(names_from = classification,
                     values_from = index) %>% 
  tidyr::separate(col = c("gene_id"), 
                  into = c("gene_id", "number"), 
                  sep = "\\.")

nmd_targets_filtered <- nmd_targets %>% 
  filter(adj_pvalue < 0.01) %>% 
  mutate(nmd = 1L)

nmd_combine <-   nmd_targets_filtered %>% 
  full_join(index_hits) %>%
  mutate_at(vars(-gene_id, -adj_pvalue, -number, -index_pval, -logfc),
            .funs = ~ if_else(is.na(.), 0L, .))


```


```{r, echo = FALSE}
nmd_sets <- nmd_combine %>% 
  select_if(is.integer) %>% 
  rename(`NMD targets` = nmd) %>% 
  as.data.frame() %>% 
  upset(nsets = 10)
nmd_sets
```

```{r save-plot, include = FALSE}
pdf(here::here("img","index_classifications", "nmd-superintronic-olaps.pdf"))
print(nmd_sets)
dev.off()
```

```{r}
ans <- nmd_combine  %>% 
  filter(nmd == 1) %>%
  tidyr::pivot_longer(cols = c(-gene_id, -adj_pvalue, -number, -index_pval, -logfc,-nmd),
                      names_to = "classification") %>% 
  group_by(classification) %>% 
  summarise(count = sum(value)) %>% 
  inner_join(count(res, classification)) %>% 
  tidyr::separate(classification, into = c("classification", "other")) %>% 
  select(-other) %>% 
  group_by(classification) %>% 
  summarise(nmd_count = sum(count),total_count = sum(n)) %>% 
  mutate(proportion = nmd_count / total_count,
         classification = if_else(nchar(classification) == 0, "+/-", classification)) %>% 
  arrange(desc(proportion))
ans
```

## Overlaps with intronic split reads


We have a DGEList containing split reads over intronic parts of a gene
for the total RNA samples, we show how many genes have greater than 10
intron split reads on average accross samples within a gene:

```{r read-splits, echo = FALSE}
sr_dge <- readRDS(here::here("data", "split_reads_intron_dgelist.rds"))

split_counts <- sr_dge$counts

totals <- tibble::tibble(
  gene_id = rownames(split_counts),
  count = unname(rowMeans(split_counts))
) %>% 
  group_by(gene_id) %>% 
  summarise(total_over_ten = sum(count >= 10)) %>% 
  mutate(over_ten = as.integer(total_over_ten > 0))
count(totals, over_ten)
```

```{r split-totals}
split_set <- totals %>% 
  full_join(dex_set) %>% 
  select(gene_id, over_ten, index, `superintronic:HCC287_total-RNA`) %>% 
  mutate_at(vars(-gene_id), 
            .funs = ~ if_else(is.na(.), 0L, .))

intron_split <- upset(as.data.frame(split_set))
intron_split
```


```{r split-plot, include = FALSE}
pdf(here::here("img","index_classifications", "split-reads-index-superintronic-olaps.pdf"))
print(intron_split)
dev.off()
```
