---
title: "Overlaps with index classifications"
author: "Stuart Lee"
date: "`r Sys.Date()`"
output: 
  bookdown::html_document2:
    toc: true
    number_sections: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      message = FALSE,
                      warning = FALSE,
                      fig.path = here::here("img/"))
library(plyranges)
parts <- readRDS(here::here("data", "complete-annotation.rds"))

hits <- readRDS(here::here("data", "superintronic-hits.rds"))

parts_tbl <- parts %>% 
  select(gene_id, gene_name, .drop_ranges = TRUE) %>% 
  tibble::as_tibble()

source(here::here("R/prettycov.R"))
```

## Extract superintronic hits 


* superintronic run on all kits in each of the HCC827 and NCIH1975 celllines
```{r si-sets}
# only polyA results
library(dplyr)
polyA_sets <- hits %>%
  filter(Kit == "polyA") %>% 
  distinct(gene_id) %>%
  arrange(gene_id) %>% 
  mutate(value = 1, 
         var = paste0("superintronic:", gsub("-", "", CellLine),"_", Kit)) %>% 
  ungroup() %>%  
  plyranges::select(gene_id, var, value) %>% 
  tidyr::spread(var, value) %>% 
  dplyr::mutate_at(2:3, .funs = ~ dplyr::if_else(is.na(.), 0L, 1L)) %>% 
  left_join(parts_tbl)

totalRNA_sets <- hits %>%
  filter(Kit == "total-RNA") %>% 
  distinct(gene_id) %>%
  arrange(gene_id) %>% 
  mutate(value = 1, 
         var = paste0("superintronic:", gsub("-", "", CellLine),"_", Kit)) %>% 
  ungroup() %>%  
  plyranges::select(gene_id, var, value) %>% 
  tidyr::spread(var, value) %>% 
  dplyr::mutate_at(2:3, .funs = ~ dplyr::if_else(is.na(.), 0L, 1L)) %>% 
  left_join(parts_tbl)
```


## Run index

```{r}
library(index)
exon <- readRDS(system.file("extdata/exon_dge.Rds", package = "index"))
intron <- readRDS(system.file("extdata/intron_dge.Rds", package = "index"))
group <- readRDS(system.file("extdata/group.Rds", package = "index"))

x <- index_analysis(exon, intron, group)

res <- x$tops$intron %>% 
  mutate(classification = x$category) %>% 
  as_tibble()

index_res <- res %>% 
  filter(adj.P.Val < 0.01) %>% 
  select(gene_id = GeneID, index_pval = adj.P.Val, logfc = logFC) %>% 
  mutate(`index` = 1L) 

index_res
```

Using index, we specify a cut-off on differential intron counts at an 
FDR of 0.01. 


## index sets overlapping total RNA comparisons

All hits discovered by superintronic except for 19 genes are 
obtained by the index analysis using the beforementioned FDR threshold. 

```{r index_olap}
library(UpSetR)
dex_set <- index_res %>% 
  full_join(totalRNA_sets) %>% 
  mutate_at(vars(index, starts_with("superintronic")), 
            .funs = ~ if_else(is.na(.), 0L, .)) 

dex_upset <- upset(as.data.frame(dex_set))
dex_upset
```

```{r index-upset-1, include = FALSE}
pdf(here::here("img", "index-superintronic-olaps-fdr_0.01.pdf"))
print(dex_upset)
dev.off()
```

How many genes in DEList overlap with HCC287 total RNA = 13 
How many genes in DEList overlap with NCIH1975 total RNA = 16
How many genes overlap between the overlapping genes = 12.  

If we look at the direction of the log fold-change  when there is overlap, the
superintronic results are concordant (the  index run has contrasts where
upregulated is comparing NCIH1975 to HCC827)

```{r}
dex_set %>% 
  count(`superintronic:HCC287_total-RNA`, 
        `superintronic:NCIH1975_total-RNA`, 
        index_sign = sign(logfc)) %>% 
  filter(`superintronic:HCC287_total-RNA` != 0 | 
           `superintronic:NCIH1975_total-RNA` != 0,
         !is.na(index_sign))
```

## Coverage plots for intron +/-

Next we look at the index category for `intron+-`, we first produce coverage
plots for the top 10 genes, ordered by t-statistic values. We do this over the
full design, instead of the just the total-RNA samples.

```{r intron-plus-minus}
# sort by t-statistic
intron_plus_minus <- res %>% 
  filter(grepl("Intron", classification), adj.P.Val < 0.01) %>%
  arrange(desc(abs(t)))

top10_genes <- intron_plus_minus %>% 
  dplyr::slice(1:10) %>% 
  select(gene_id =  GeneID, classification) %>% 
  mutate_all(as.character)

top10_genes
```

```{r read-rds, echo = FALSE}
cvg_over_features <- readRDS(here::here("data","complete-coverage.rds")) 
design <- readRDS(here::here("data", "design.rds"))
```

```{r cov-plots, warning = FALSE, echo = FALSE, message = FALSE, fig.width = 8}
library(ggplot2)
for (i in seq_len(nrow(top10_genes))) {
  current <-  top10_genes[i,]
  track_plot <- pretty_cov_plot(cvg_over_features, 
                                parts,
                                current,
                                design,
                                heights = c(4, 1/4))
  print(track_plot)
  out_name <- paste0(current$gene_id, "-", 
                     tolower(current$classification),
                     "_index_coverage.pdf")
  ggsave(here::here("img", out_name), 
         track_plot,
         width = 8, 
         units = "in")
}
```



## Overlaps with index classifications

Next we look at the overlaps between superintronic results
and all the  index classifications. Wen see there is only
gene classified by superintronic, that doesn't come up as a hit
in any of the index categories.  

```{r classification-set, echo = FALSE}
complete_set <- res %>%
  select(gene_id = GeneID, 
         index_pval = adj.P.Val, 
         logfc = logFC, 
         classification) %>% 
  filter(nchar(classification) > 0) %>% 
  mutate(index = 1L) %>% 
  tidyr::pivot_wider(names_from = classification,
                     values_from = index) %>%
  full_join(totalRNA_sets) %>%
  # rename(`HCC287 total RNA` = `superintronic:HCC287_total-RNA`,
  #        `NCIH1975 total RNA` = `superintronic:NCIH1975_total-RNA`) %>% 
  mutate_at(vars(-gene_id:-logfc, -gene_name),
            .funs = ~ if_else(is.na(.), 0L, .))

all_classification_upset <- upset(as.data.frame(complete_set),
                                  nsets = 10,
                                  text.scale = 0.7)
all_classification_upset
```

The remaining results make sense, most superintronic results fall into either `+` or `-` categories, or the intron `+/-` categories, while
a few fall into the remaining categories.

```{r index-upset-2, include = FALSE}
pdf(here::here("img", "index-categories-superintronic-olaps.pdf"))
print(all_classification_upset)
dev.off()
```


Below we show a table, with the proportion of superintronic hits
for each CellLine within each category. In the tables, missing
values indicate the result was unique to superintronic.

These are the proportions for HCC287 cellline hits:

```{r, echo = FALSE}
by_category <- res %>%
  select(gene_id = GeneID, 
         index_pval = adj.P.Val, 
         logfc = logFC, 
         classification) %>% 
  filter(nchar(classification) > 0) %>% 
    full_join(totalRNA_sets) %>%
  mutate_at(vars(-gene_id:-classification, -gene_name),
            .funs = ~ if_else(is.na(.), 0L, .)) %>% 
  group_by(classification)

by_category %>% 
  filter(`superintronic:HCC287_total-RNA` == 1) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(proportion = n / sum(n)) %>% 
  arrange(desc(proportion))
```
We see that again the results make sense, where most classifications
fall into minus categories.

```{r, echo = FALSE}
by_category %>% 
  filter(`superintronic:NCIH1975_total-RNA` == 1) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(proportion = n / sum(n)) %>% 
  arrange(desc(proportion))
```

Likewise, the proportions for NCIH1975 hits, show that most hits
fall into the plus category, which makes sense. 
